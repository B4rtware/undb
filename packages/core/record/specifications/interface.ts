import type { CompositeSpecification, ISpecVisitor } from '@egodb/domain'
import { type ISpecification } from '@egodb/domain'
import { type Record } from '../record'
import type { BoolIsFalse, BoolIsTrue } from './bool.specification'
import type { DateRangeEqual } from './date-range.specification'
import type {
  DateEqual,
  DateGreaterThan,
  DateGreaterThanOrEqual,
  DateIsToday,
  DateLessThan,
  DateLessThanOrEqual,
} from './date.specification'
import type { NullSpecification } from './null.specification'
import type {
  NumberEqual,
  NumberGreaterThan,
  NumberGreaterThanOrEqual,
  NumberLessThan,
  NumberLessThanOrEqual,
} from './number.specification'
import type { WithRecordCreatedAt } from './record-created-at.specification'
import type { WithRecordId } from './record-id.specifaction'
import type { WithRecordTableId } from './record-table-id.specification'
import type { WithRecordUpdatedAt } from './record-updated-at.specification'
import type { WithRecordValues } from './record-values.specification'
import type { ReferenceEqual } from './reference.specification'
import type { SelectEqual, SelectIn } from './select.specification'
import type { StringContain, StringEndsWith, StringEqual, StringRegex, StringStartsWith } from './string.specification'

interface IRecordSpecVisitor {
  idEqual(s: WithRecordId): void
  tableIdEqual(s: WithRecordTableId): void

  createdAt(s: WithRecordCreatedAt): void
  updatedAt(s: WithRecordUpdatedAt): void

  values(s: WithRecordValues): void
}

interface IRecordValueVisitor {
  stringEqual(s: StringEqual): void
  stringContain(s: StringContain): void
  stringStartsWith(s: StringStartsWith): void
  stringEndsWith(s: StringEndsWith): void
  stringRegex(s: StringRegex): void

  numberEqual(s: NumberEqual): void
  numberGreaterThan(s: NumberGreaterThan): void
  numberLessThan(s: NumberLessThan): void
  numberGreaterThanOrEqual(s: NumberGreaterThanOrEqual): void
  numberLessThanOrEqual(s: NumberLessThanOrEqual): void

  dateEqual(s: DateEqual): void
  dateGreaterThan(s: DateGreaterThan): void
  dateLessThan(s: DateLessThan): void
  dateGreaterThanOrEqual(s: DateGreaterThanOrEqual): void
  dateLessThanOrEqual(s: DateLessThanOrEqual): void
  dateIsToday(s: DateIsToday): void

  null(s: NullSpecification): void
  dateRangeEqual(s: DateRangeEqual): void

  selectEqual(s: SelectEqual): void
  selectIn(s: SelectIn): void

  boolIsTrue(s: BoolIsTrue): void
  boolIsFalse(s: BoolIsFalse): void

  referenceEqual(s: ReferenceEqual): void
}

export type RecordCompositeSpecification = CompositeSpecification<Record, IRecordVisitor>

export type IRecordSpec = ISpecification<Record, IRecordVisitor>

export type IRecordVisitor = IRecordSpecVisitor & IRecordValueVisitor & ISpecVisitor
