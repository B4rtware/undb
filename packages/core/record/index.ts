export * from './fixtures'
export * from './record'
export * from './record-tree-query-model'
export * from './record.factory'
export * from './record.query-model'
export * from './record.repository'
export * from './record.schema'
export * from './record.type'
export * from './specifications'
export * from './value-objects'
