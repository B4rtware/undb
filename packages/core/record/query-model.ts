import type { Option } from 'oxide.ts'
import type { IQueryRecordSchema } from './record.type'
import type { IRecordSpec } from './specifications'

export interface IRecordQueryModel {
  findOne(spec: IRecordSpec): Promise<Option<IQueryRecordSchema>>
  findOneById(id: string): Promise<Option<IQueryRecordSchema>>
  find(): Promise<IQueryRecordSchema[]>
}
