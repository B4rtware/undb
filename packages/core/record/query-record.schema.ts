export interface QueryRecord {
  id: string
}
