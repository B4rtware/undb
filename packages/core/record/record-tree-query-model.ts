import type { TreeField } from '../field'
import type { TableSchemaIdMap } from '../value-objects'
import type { IQueryTreeRecords } from './record.type'
import type { IRecordSpec } from './specifications'

export interface IRecordTreeQueryModel {
  /**
   * returns root categories with sub categories inside
   * @param tableId - tableId
   * @param field - tree field
   * @param spec - spec
   * @param schema - schema
   */
  findTrees(tableId: string, field: TreeField, spec: IRecordSpec, schema: TableSchemaIdMap): Promise<IQueryTreeRecords>
}
