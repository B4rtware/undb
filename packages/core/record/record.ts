import type { TableId } from '../value-objects'
import { RecordId } from './value-objects'
import { RecordValues } from './value-objects/record-values.vo'

export class Record {
  public id: RecordId = RecordId.create()
  public tableId!: TableId
  public values: RecordValues = RecordValues.empty()
}
