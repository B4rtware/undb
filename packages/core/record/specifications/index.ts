export type { IRecordSpec, IRecordSpecVisitor } from './interface'
export * from './record-id.specifaction'
export * from './record-table-id.specification'
export * from './string.specification'
