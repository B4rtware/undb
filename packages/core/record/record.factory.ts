import { and } from '@egodb/domain'
import type { Result } from 'oxide.ts'
import type { TableSchemaIdMap } from '../value-objects'
import { Record } from './record'
import type { IQueryRecordSchema, Records } from './record.type'
import { WithRecordCreatedAt, WithRecordId, WithRecordTableId, WithRecordValues } from './specifications'
import type { RecordCompositeSpecification } from './specifications/interface'

export class RecordFactory {
  static create(...specs: RecordCompositeSpecification[]): Result<Record, string>
  static create(spec: RecordCompositeSpecification): Result<Record, string>

  static create(spec: RecordCompositeSpecification | RecordCompositeSpecification[]): Result<Record, string> {
    if (Array.isArray(spec)) {
      return and(...spec)
        .unwrap()
        .mutate(Record.create())
    }
    return spec.mutate(Record.create())
  }

  static fromQueryRecords(rs: IQueryRecordSchema[], schema: TableSchemaIdMap): Records {
    return rs.map((r) => this.fromQuery(r, schema).unwrap())
  }

  static fromQuery(r: IQueryRecordSchema, schema: TableSchemaIdMap): Result<Record, string> {
    return this.create(
      WithRecordId.fromString(r.id)
        .and(WithRecordTableId.fromString(r.tableId).unwrap())
        .and(WithRecordCreatedAt.fromDate(r.createdAt))
        .and(WithRecordValues.fromObject(schema, r.values)),
    )
  }
}
