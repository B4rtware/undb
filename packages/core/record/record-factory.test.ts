import { RecordFactory } from './record.factory'
import { WithRecordId, WithRecordTableId } from './specifications'
import { RecordCompositeSpecification } from './specifications/interface'

test.each<RecordCompositeSpecification>([
  WithRecordTableId.fromString('table').unwrap().and(WithRecordId.fromString('record')),
])('should create record', (spec) => {
  const record = RecordFactory.create(spec)

  expect(record).toMatchSnapshot()
})
