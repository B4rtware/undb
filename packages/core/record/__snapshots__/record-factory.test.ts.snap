// Vitest Snapshot v1

exports[`should create record 1`] = `
ResultType {
  Symbol(Val): Record {
    "id": RecordId {
      "props": {
        "value": "record",
      },
    },
    "tableId": TableId {
      "props": {
        "value": "table",
      },
    },
    "values": RecordValues {
      "props": Map {},
    },
  },
  Symbol(T): true,
}
`;

exports[`should create record 2`] = `
ResultType {
  Symbol(Val): Record {
    "id": RecordId {
      "props": {
        "value": "record",
      },
    },
    "tableId": TableId {
      "props": {
        "value": "table",
      },
    },
    "values": RecordValues {
      "props": Map {},
    },
  },
  Symbol(T): true,
}
`;
